plat/common/lcpu.c

```diff
@@ -48,3 +165,424 @@ void ukplat_lcpu_halt_to(__nsec until)
 	time_block_until(until);
 	ukplat_lcpu_restore_irqf(flags);
 }
+
+__lcpuid ukplat_lcpu_id(void)
+{
+	return lcpu_arch_id();
+}
+
+#ifdef CONFIG_HAVE_SMP
+#ifdef CONFIG_UKPLAT_LCPU_IDISIDX
+/* For many VMMs the CPU ID is simply a sequentially increasing number with
+ * the BSP always at index 0. This is essentially the definition of __lcpuidx
+ * so in this case, we can simply use the ID as IDX.
+ */
+__lcpuidx ukplat_lcpu_idx(void)
+{
+	__lcpuidx this_cpu_idx = (__lcpuidx) ukplat_lcpu_id();
+
+	UK_ASSERT(this_cpu_idx < lcpu_count);
+	UK_ASSERT(lcpus[this_cpu_idx].idx == this_cpu_idx);
+	return this_cpu_idx;
+}
+#else /* CONFIG_UKPLAT_LCPU_IDISIDX */
+__lcpuidx ukplat_lcpu_idx(void)
+{
+	__lcpuid this_cpu_id  = ukplat_lcpu_id();
+	__u32 i, this_cpu_idx = (__u32) (-1);
+
+	/** TODO: Until we have a better way just do a linear search */
+	for (i = 0; i < lcpu_count; i++)
+		if (lcpus[i].id == this_cpu_id) {
+			UK_ASSERT(lcpus[i].idx == (__lcpuidx) i);
+			this_cpu_idx = i;
+			break;
+		}
+
+	UK_ASSERT(this_cpu_idx != (__u32) (-1));
+
+	return (__lcpuidx) this_cpu_idx;
+}
+#endif /* CONFIG_UKPLAT_LCPU_IDISIDX */
+
+int lcpu_fn_enqueue(struct lcpu *lcpu, const struct ukplat_lcpu_func *fn)
+{
+	void (*old_fn)(struct __regs *, const struct ukplat_lcpu_func *);
+
+	UK_ASSERT(fn->fn);
+
+	old_fn = ukarch_load_n(&lcpu->fn.fn);
+
+	/* Check if the slot is empty */
+	if (old_fn != NULL)
+		return -EAGAIN;
+
+	/* It is empty, try to store the function */
+	if (ukarch_compare_exchange_sync(&lcpu->fn.fn, old_fn,
+					 fn->fn) != fn->fn)
+		return -EAGAIN;
+
+	/* We have acquired the slot! Also store the user argument.
+	 * It is safe to do it afterwards, because the RUN IRQ handler will
+	 * only take one function and return afterwards. And we only raise the
+	 * IRQ after finishing setup.
+	 */
+	lcpu->fn.user = fn->user;
+
+	/* Ensure everything is written back when we return and the arch
+	 * support code will raise the IRQ
+	 */
+	wmb();
+
+	return 0;
+}
+
+static void lcpu_fn_dequeue(struct lcpu *this_lcpu, struct ukplat_lcpu_func *fn)
+{
+	*fn = this_lcpu->fn;
+
+	/* Ensure that we have captured the whole function object */
```

Comment:
I understand the purpose of this barrier, but the "whole function object" statement confused me, as it implies... an object rather than a pointer. Or I think I understand but I'm actually missing something? 🤔

Reply:
This is not only a single pointer, but a struct consisting of a function pointer and a user argument. Thus the term function object. Do you think this is still confusing then?

Reply:
No it was just me not being aware of the copy struct by pointer trick. In fact the comment is useful. Live and learn 😊